src/security/jdbc/src/main/java/org/geoserver/security/jdbc/JDBCRoleStore.java

```diff
@@ -82,6 +86,12 @@ protected void addRoleProperties(GeoServerRole role, Connection con)
                 ps.setObject(3, propertyVal);
                 ps.execute();
             }
+        } catch (SQLException ex) {
+            if (ex.getMessage().contains("duplicate key value violates unique constraint")) {
```

Comment:
So... these kind of messages are very database and jdbc driver specific and this seems to cater only for PostgreSQL

Reply:
Thanks for the valuable input @mprins.  Would it be acceptable if I checked for different messages for the following RDBMSs?

| **RDBMS**        | **Error Message**                                                                                                                                                                 |
|------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| **SQL Server**   | `Msg 2627, Level 14, State 1, Line 1 Violation of UNIQUE KEY constraint 'constraint_name'. Cannot insert duplicate key in object 'object_name'. The duplicate key value is (duplicate_value).` |
| **MySQL**        | `ERROR 1062 (23000): Duplicate entry 'duplicate_value' for key 'constraint_name'`                                                                                                |
| **Oracle**       | `ORA-00001: unique constraint (constraint_name) violated`                                                                                                                        |
| **SQLite**       | `SQLITE_CONSTRAINT_UNIQUE: UNIQUE constraint failed: table_name.column_name`                                                                                                      |
| **PostgreSQL**   | `ERROR: duplicate key value violates unique constraint "constraint_name" DETAIL: Key (column_name)=(duplicate_value) already exists.`                                            |
| **DB2**          | `SQL0803N  One or more values in the INSERT statement, UPDATE statement, or foreign key update caused by a DELETE statement are not valid because the primary key, unique constraint or unique index identified by "constraint_name" constrains table "table_name" from having duplicate values for the index key.` |
| **H2**           | `Unique index or primary key violation: "constraint_name index_id"; SQL statement: ...`                                                                                          |

Reply:
Pheraps you want to catch a [SQLIntegrityConstraintViolationException](https://docs.oracle.com/javase/8/docs/api/java/sql/SQLIntegrityConstraintViolationException.html) instead, and be done with it?

However, I would take a step back, before getting bogged down into details. What are the XML services doing when there is a duplicate instead? What are the consequences of ignoring the exception?

When I see how this is presented, it seems like a "I want to see this exception gone" but I'd like to see a bit of higher-level discussion instead.  E.g. complaining when inserting a group that already exists does not sound like a bad idea, would it be better to align the XML modules instead? Does the client have a quick way to check for duplicates before entering one?

Reply:
Added some comments to the jira ticket and examples of why trying to catch [`SQLIntegrityConstraintViolationException`](https://docs.oracle.com/javase/8/docs/api/java/sql/SQLIntegrityConstraintViolationException.html) will not work in general

Reply:
Q: Where should discussions take place, in GitHub PRs or in Jira tickets?  Probably not both ;-)

> What are the XML services doing when there is a duplicate instead? What are the consequences of ignoring the exception?

The current (XML service) implementation considers a duplicate a no-op and returns 200.  I personally think that is correct, as the outcome was achieved, but if you want me to follow up this PR with another that changes them all to 409 Conflict, I can look at that.

> Does the client have a quick way to check for duplicates before entering one?

[GET](https://docs.geoserver.org/2.25.x/en/user/rest/api/userrole.html#rest-usergroup-service-servicename-user-user-groups) all users for a specified group (or the other way around) and check that the response contains the provided user is easy enough to do, but it's not as easy as handling a 409 Conflict error.

Reply:
Higher-level discussions are best placed on list, it's hard to get everyone's attention on the PR comments.
Gut feeling, if someone is duplicating something existing, a 409 seems like a better choice, but this is exactly the kind of decision where having all consider the issue is best.

Reply:
Feedback via PSC meeting: 

check first: if association is already there then return success  
Avoids the unmaintainable mess above.